common-solidity/coordination-contract/contracts/CrosschainCoordinationV1.sol

```diff
@@ -473,11 +508,148 @@ contract CrosschainCoordinationV1 is CrosschainCoordinationInterface {
+        for (uint i = 0; i < sidechainsNoOfKeys; i++) {
+            publicKeyArray[i] = sidechains[_sidechainId].publicKeys[i];
+        }
+        return(publicKeyArray);
+    }
+
+    /**
+     * Get the last active key for the sidechain
+     *
+     * @param _sidechainId The 256 bit sidechain identifier to which this public key belongs
+     * @return _activeKey latest active public key
+     */
+    function getLastActivePublicKey(uint256 _sidechainId) private view returns ( uint _versionNumber, uint256 _status, uint _blockNumber, bytes memory _key){
+        // If there is exactly one public key then this is the one assigned when the sidechain was created & we would expect it to have
+        // a status of active, but this probably doesn't matter...
+        PublicKey [] memory publicKeyArray = getPublicKeyArray(_sidechainId);
+        uint256 sidechainsNoOfKeys = sidechains[_sidechainId].numberOfKeys;
+        bool active;     // default initialisation is FALSE
+        uint i;
+        for (i = sidechainsNoOfKeys; (i == 0 || active); i--) {
+            publicKeyArray[i-1] = sidechains[_sidechainId].publicKeys[i];
+            if (publicKeyArray[i-1].status == uint256(PublicKeyStatus.KEY_CURRENT)) {
+                active = true;
+            }
+        }
+        return(publicKeyArray[i].versionNumber, publicKeyArray[i].status, publicKeyArray[i].blockNumber, publicKeyArray[i].key);
+    }
+
+    /**
+     * Set the Sidechain's public key, version, status & block number
+     *
+     * @param  _sidechainId    The 256 bit sidechain identifier to which this public key belongs
+     * @param  _publicKey      The new public key for the sidechain
+     */
+ //   function setPublicKey(uint256 _sidechainId, uint _versionNumber, uint256 _status, bytes storage _publicKey) private {
+    function setPublicKey(uint256 _sidechainId, uint _versionNumber, uint256 _status, bytes memory _publicKey) private {
+       // Check if we are adding or replacing a public key entry
+        uint256 index = sidechains[_sidechainId].numberOfKeys;
+        // Need to replace entry and not add an entry - assume that we are updating the block number too
+        // TODO - Do we need to update the block number?
+        if (index == _versionNumber){
+            sidechains[_sidechainId].publicKeys[index-1].status = _status;
+            sidechains[_sidechainId].publicKeys[index-1].blockNumber = block.number;
+        } else {
+            sidechains[_sidechainId].publicKeys[index].versionNumber = _versionNumber;
+            sidechains[_sidechainId].publicKeys[index].status = _status;
+            sidechains[_sidechainId].publicKeys[index].blockNumber = block.number;
+            sidechains[_sidechainId].publicKeys[index].key = _publicKey;
+            sidechains[_sidechainId].numberOfKeys = index + 1;
+        }
+    }
+
+    /**
+        * There was a majority vote to change the public key of the Sidechain, so change the public key as has been voted on
```

Comment:
The voting algorithm may not have been a majority vote. At this point, all we know is that the voting algorithm indicates the vote succeeded and should be actioned.

Reply:
Deleted the word 'majority'